impl_core/src/main/java/io/opencensus/implcore/trace/export/SampledSpanStoreImpl.java

```diff
@@ -48,6 +49,10 @@
       NUM_SAMPLES_PER_LATENCY_BUCKET * NUM_LATENCY_BUCKETS
           + NUM_SAMPLES_PER_ERROR_BUCKET * NUM_ERROR_BUCKETS;
 
+  // Used to stream the register/unregister events to the implementation to avoid lock contention
+  // between the main threads and the worker thread.
+  private final EventQueue eventQueue;
```

Comment:
It doesn't prevent the contention on `samples`. There is always a contention if `considerForSampling` ever has a `synchronized` in the main path.

Reply:
The way it works is the following:

Critical path Thread -> EventQueues(Disruptor) -> Worker Thread -> SampledSpanStoreImpl.lock

As you can see the critical path Thread does not touch the lock in the SampledSpanStoreImpl only the Worker thread interacts with that mutex.

For the end event is the same because the considerForSampling is also recorded as part of the SpanEndEvent not directly applied to the SampledSpanStoreImpl.

Reply:
I see. You are already streaming `considerForSampling()` using the EventQueue. I misunderstood.

Reply:
For the moment we have one thread. We will work later to improve and support multiple threads then your suggestion of using thread safe data structures will make a lot of sense.

Also currently getRegisteredSpanNamesForCollection is a place where the critical path can interact with the lock, but we marked that as test only (because indeed it is used only in tests).